pennylane/interfaces/batch/jax.py

```diff
@@ -0,0 +1,257 @@
+    params,
+    tapes=None,
+    device=None,
+    execute_fn=None,
+    gradient_fn=None,
+    gradient_kwargs=None,
+    _n=1,
+):  # pylint: disable=dangerous-default-value,unused-argument
+
+    # Only have scalar outputs
+    total_size = len(tapes)
+    total_params = np.sum([len(p) for p in params])
+
+    @jax.custom_vjp
+    def wrapped_exec(params):
+        def wrapper(p):
+            """Compute the forward pass."""
+            new_tapes = []
+
+            for t, a in zip(tapes, p):
+                new_tapes.append(t.copy(copy_operations=True))
+                new_tapes[-1].set_parameters(a)
+
+            with qml.tape.Unwrap(*new_tapes):
+                res, _ = execute_fn(new_tapes, **gradient_kwargs)
+
+            return np.stack(res)
+
+        res = host_callback.call(
+            wrapper, params, result_shape=jax.ShapeDtypeStruct((total_size, 1), dtype)
+        )
+        return res
+
+    def wrapped_exec_fwd(params):
+        return wrapped_exec(params), params
+
+    def wrapped_exec_bwd(params, g):
+
+        if isinstance(gradient_fn, qml.gradients.gradient_transform):
+
+            def non_diff_wrapper(args):
+                """Compute the VJP in a non-differentiable manner."""
+                new_tapes = []
+                p = args[:-1]
+                dy = args[-1]
+
+                for t, a in zip(tapes, p):
+                    new_tapes.append(t.copy(copy_operations=True))
+                    new_tapes[-1].set_parameters(a)
+                    new_tapes[-1].trainable_params = t.trainable_params
+
+                vjp_tapes, processing_fn = qml.gradients.batch_vjp(
+                    new_tapes,
+                    dy,
+                    gradient_fn,
+                    reduction="append",
+                    gradient_kwargs=gradient_kwargs,
+                )
+
+                partial_res = execute_fn(vjp_tapes)[0]
+                res = processing_fn(partial_res)
+                return np.concatenate(res)
+
+            args = tuple(params) + (g,)
+            vjps = host_callback.call(
+                non_diff_wrapper,
+                args,
+                result_shape=jax.ShapeDtypeStruct((total_params,), dtype),
+            )
+
+            param_idx = 0
+            res = []
+
+            # Group the vjps based on the parameters of the tapes
+            for p in params:
+                res.append(vjps[param_idx : param_idx + len(p)])
+                param_idx += len(p)
+
+                # Unpack if we have the vjp for a single parameter
+                if len(p) == 1:
+                    res[-1] = res[-1][0]
+
+            # Unwrap partial results into arrays if need be
+            need_unwrapping = any(r.ndim != 0 for r in res)
+            if need_unwrapping:
+                unwrapped_res = []
+                for r in res:
+                    if r.ndim != 0:
+                        r = [jnp.array(p) for p in r]
+                    unwrapped_res.append(r)

```

Comment:
I'm also not 100% following this code block here 🤔

Reply:
This maps multi-dimensional arrays into groups of scalar arrays:
`[DeviceArray([-0.9553365], dtype=float32), DeviceArray([0., 0.], dtype=float32)]`
is mapped to
`[[DeviceArray(-0.9553365, dtype=float32)], [DeviceArray(0., dtype=float32), DeviceArray(0., dtype=float32)]]`.

This is in response to an error I've encountered from JAX otherwise:
```python
The above exception was the direct cause of the following exception:
tests/interfaces/test_batch_jax.py:53: in test_jacobian_options
    res = jax.grad(cost)(a, device=dev)
../../anaconda3/lib/python3.8/site-packages/jax/_src/lax/lax.py:4734: in _scatter_shape_rule
    raise TypeError(f"Updates tensor must be of rank {expected_updates_rank}; "
E   TypeError: Updates tensor must be of rank 0; got 1.
```
This does differ from how the `autograd` interface works, however it seems like JAX requires a different return type for backward to compute the gradient.

Reply:
got it!

Reply:
@antalszava your comment above was really useful, might be worth incorporating it into a comment block in the code?

Reply:
Added the info about the mapping including the example.